actually process `<constraintDecl>`

sydb opened the pull request:
[Note: There is no issue married to this PR.]

Wow. I first found a teeny bug in extract-isosch.xsl which occurred in the case of more than one prefix bound to a particular namespace URI. A warning message told the user that only the 1st prefix would be used, but the code tried to use all of them, and thus failed kinda dramatically.

But while fixing & testing this _tiny_ bug fix (only changes 1 line of code), I realized that our entire processing chain pretty much ignored a `<constraintDecl>` that is in the base (or source) ODD.

So I added code to odd2odd.xsl that will (I hope) merge both the `<constraintDecl>`s from the customization ODD with those from the base ODD on a scheme-by-scheme basis, putting the result in the `<encodingDesc>` of the output compiled ODD. I then added code to extract-isosch.xsl to pay attention to `<constraintDecl>`s wherever they may be.

This code passes both Test/ and Test2/, and I was able to build the WWP schema using this branch. (Which says something, because it is a big customization that has a lot of Schematron.)

sydb committed:
Stylesheets group figured out
how to process constraintDecl when processing the Schematron that
goes into RELAX NG output. Still no tests.

sydb committed:
Stylesheets group figured out
how to process constraintDecl when processing the Schematron that
goes into RELAX NG output. Still no tests.

sydb committed:
Bug fix: the warning message very
clearly says that if there is more than one declared
namespace with the same URI, only the prefix of the 1st
is used. But the code didn’t actually **do** that. It
tried to use both, and thus failed miserably.

sydb committed:
Big changes:
Merge constraintDecls from both customization ODD and base ODD
into compiled output ODD, and extract from the resulting
constraintDecls.

joeytakeda commented on odds/odd2odd.xsl:
```diff
@@ -534,6 +527,43 @@ of this software, even if advised of the possibility of such damage.
+		  select="( document(tei:workOutSource(.))//tei:constraintDecl, //tei:constraintDecl )"/>
+    <xsl:if test="child::tei:encodingDesc or count( $constraintDecls ) > 0">
+      <tei:encodingDesc>
+	<xsl:apply-templates select="tei:encodingDesc/@*" mode="#current"/>
```
It doesn't look like other `tei:encodingDesc` children are being processed here (e.g. `listPrefixDef` etc). I'm presuming that other encodingDesc children should be included here too, so should it be something like:

```xml
<xsl:apply-templates select="tei:encodingDesc/node() except //tei:constraintDecl" mode="#current"/>
```

sydb replied:
Oh wow. Good point. (And fast — only ½ hour.) I am not sure how that got dropped. (But I used 
~~~xml
<xsl:apply-templates select="tei:encodingDesc/node() except tei:constraintDecl" mode="#current"/>
~~~
… does the leading `//` matter there?)

joeytakeda commented on odds/odd2odd.xsl:
```diff
@@ -534,6 +527,43 @@ of this software, even if advised of the possibility of such damage.
+	<xsl:variable name="constraintDecl_schemes" select="$constraintDecls/@scheme" as="xs:string*"/>
+	<xsl:for-each select="distinct-values( $constraintDecl_schemes )">
+	  <constraintDecl scheme="{.}">
+            <xsl:apply-templates select="key('odd2odd-CONSTRAINTDECLS-by-SCHEME', ., document(tei:workOutSource( $top )) )"/>
```
Unless I'm missing something, the `key('odd2odd-CONSTRAINTDECLS-by-SCHEME')` returns a `constraintDecl`, so wouldn't this end up duplicating the wrapping `constraintDecl`? E.g. should this be:

```xml
<xsl:apply-templates
 select="key('odd2odd-CONSTRAINTDECLS-by-SCHEME', ., document(tei:workOutSource( $top )) )/node()" 
mode="#current"/>
```

sydb replied:
Uhhh … maybe. Makes sense, but I would want to actually test it to see for sure, which I probably can’t do for next hour or two.

sydb replied:
Yes, you are right. Perhaps worse, though, it was not returning anything because the call to tei:workOutSource() was incorrect. It could be fixed by using `$top//tei:schemaSpec` as the argument, but that makes me worry about what happens if there are more than 1 `<schemaSpec>` (crash, I expect), and how we would select one. Need to think about this a bit more. (Another possibility, of course, is to change how tei:workOutSource() works.)

sydb replied:
Whoo-hoo! I think I have a solution. Basically I just stopped using keys, and hand-assemble the list of `<constraintDecl>`s that are in force.

However, in thinking this through I think I may have stumbled upon a really bad problem. I kinda suspect that much of odd2odd.xsl is messed up, because it is processing stuff that is _not_ in the desired `<schemaSpec>`, but rather is in some other `<schemaSpec>`. (E.g. the key 'odd2odd-CHANGEATT' is set to _every_ `attDef[@mode eq 'change']`, whether it is in the `<schemaSpec>` being processed or another one.)

That is, I think, a solvable problem. The key() function can be handed the `<schemaSpec>` in question as a 3rd parameter whenever it is used. But there may be more to it than that, and even if there were not, this is not something I want to try to get done in < 48 hrs for the upcoming release.

BTW, a side note: the reason this PR works even though I have added this processing to pass0 is that `<constraintDecl>` is not allowed within `<specGrp>`.

ebeshero replied:
So, at this point what could possibly go wrong if we merge this? (Seriously, kinds of problems should we be looking for?)

sydb replied:
What could _possibly_ go wrong is a long, if not infinite, list. But at least nothing we test for in the tests will go wrong. That’s the good news. The bad news is that `<constraintDecl>` does not occur in any Stylesheets/ tests at all, and only in TEI/ tests as a side-effect of being part of the _Guidelines_. It is included in both the _Guidelines_ and one of the Exemplars in the TEI:sydb_use_constraintDecl_and_fr_CMC branch, and works there (as long as you are building against this, the Stylesheets:sydb_fix_constraintDecl_oversight_01 branch).

I think the thing(s) to test is that which we expect users to do — namely, to use `<constraintDecl scheme="schematron">` (both with and without `@xml:lang`, which should make no difference) in either the `<encodingDesc>` or as a child of `<schemaSpec>` (or both, which should generate an error) in their customization ODDs.

sydb replied:
BTW, I think the easy way to test this sort of thing is generate a customization ODD that does what you want to test, and then (using this branch) issue `bin/teitoschematron --odd /path/to/IN.odd /path/to/OUT.sch`. If the OUT.sch looks OK (you may want to compare it to the output you get from same cmd in dev branch), you may want to test that the _rest_ of processing works by using your normal method of generating RELAX NG and HTML, and spot-checking those. (But nothing I did on this branch _should_  change those outputs in any significant way, I don’t think.) Some things to test are whether `@queryBinding` makes it through, that `<sch:rule>`s make it through, and what happens if there are duplicate namespace URIs or prefixes defined in `<sch:ns>`s.

sydb replied:
Note, though, that since #697 has not been addressed, you should not expect to see `<constraintDecl>` output in the RELAX NG output.

sydb committed:
Fix omission rapidly found by
@joeytakeda — process other contents of encodingDesc, too

sydb committed:
Fix a bug that I
introduced initially (caught by @joeytakeda), but also
re-work how constraint declaration processing is handled
altogether.

sydb commented:
Well, I have found a bunch of bugs so far …

1. When `<constraintDecl>` is child of `<schemaSpec>`:
    - the section "declarations" in the output has duplicated Schematron
    - the query binding is not carried through
1. When `<constraintDecl>` is a child of `<encodingDesc>`:
   -  the namespaces declared within appear twice each in the "namespaces, declared" section
    - the section "declarations" in the output has TWICE duplicated Schematron
1. When both occur:
    - not really a problem, as this is not allowed
    - our Schematron correctly reports the error
    - that said, it would be nice if XSLT issued an error, rather than generated bogus results

Unless someone suggests a different branch is in order, I am going to start addressing these here on this branch soon. (If anyone feels inclined to step up to the plate and help, it would be greatly appreciated — but your time might be better spent on #697 instead.)

sydb committed:
Fixed some bugs, still more to go:
 * Yup, @joeytakeda’s version was right, need the “//” on the
   exception to avoid processing twice.
 * Removed entire “declarations:” block, as everything is already
   processed elsewhere.
 * Improved pattern name creation (in case there is no ancestor
   with ident=, but there is one with xml:id=, take that; if not,
   and if in constraintDecl, just say so).
 * Don’t process sch:ns elements in constraintDecl twice.
HOWEVER, I am comitting this now as I have just decided to
re-write the main "/key('CONSTRAINTs',  )" loop,
as it is giving me trouble, and probably should be an apply-
templates, anyway.

sydb committed:
More fixes. Mainly
I re-worked a large for-each to use apply-templates. Much
better. Also added code to handle queryBinding=.

sydb commented:
The commit I just (inadvertently) pushed is NOT ready for prime-time.

sydb committed:
Tiny things:
Minor bug-fix in call to makePatternID(), and re-generated
an expected result (as IDs have changed).

sydb commented:
OK, this latest version (97822677) is passing both Test/ and Test2/ in Docker. I believe all outright bugs I listed above have been fixed, but it still does not warn user if 2+ `<constraintSpec>`s with same `@scheme` in same file. (It does warn if there is a conflict in `@queryBinding` values, though.)

sydb committed:
Give warning (and
don’t process) an sch:assert or sch:report that is directly
in a constraint, without an sch:rule.

sydb committed:
Try pulling in the various fixes made to ISOSCH processing of constraintDecl recently

sydb committed:
Seem to be two commits I don’t have

martindholmes commented:
ATOP meeting tested today and this is working as intended.

martindholmes reviewed (approved):
Good to go.
